compiler/ast.nim

```diff
@@ -1039,6 +1049,86 @@ const
   defaultAlignment = -1
   defaultOffset = -1
 
+proc r*(a: TLoc): Rope {.tags: [TreeRead].} =
+  result = a.roap
+
+proc `loc=`*(p: PSym or PType; loc: TLoc) {.tags: [TreeRead, TreeWrite].} =
+  when defined(debugTLoc):
+    echo "set location"
+  assert p.location.k == loc.k or p.location.k == locNone
+  assert p.location.roap == nil or $p.location.r == $loc.r
+  system.`=`(p.location, loc)
+
+proc loc*(p: PSym or PType): TLoc {.tags: [TreeRead].} =
+  result = p.location
+
+proc mloc*(p: PSym or PType): var TLoc {.tags: [TreeRead, TreeWrite].} =
+  result = p.location
+  when defined(debugTLoc):
+    echo "mut loc"
+
+proc setLocation*(p: PSym or PType; a: TLoc) {.tags: [TreeWrite, TreeSafe].} =
+  ## this should be run almost nowhere
+  p.location = a
+
+proc mr*(a: var TLoc): var Rope {.tags: [TreeRead, TreeWrite].} =
+  when defined(debugTLoc):
+    echo "get rope mut"
+  result = a.roap
+
+proc clearRope*(a: TLoc) {.tags: [TreeWrite].} =
+  assert a.roap == nil
+  when defined(debugTLoc):
+    echo "clear imm"
+
+proc clearRope*(a: var TLoc) {.tags: [TreeWrite].} =
+  a.roap = nil
+  when defined(debugTLoc):
+    echo "clear mut"
+
+proc setRope*(a: TLoc; roap: Rope) {.tags: [TreeWrite].} =
+  ## a trap to catch bad attempts to mutate an immutable
```

Comment:
I don't understand; why not catch at CT by removing this proc?

Reply:
The proc conveys the semantics we want, but it shouldn't necessarily fail in cases where we are gratuitously performing a set.

Anyway, it's likely deadcode after the AST becomes immutable.  This is an older impl.